actors/builtin/power/power_actor.go

```diff
@@ -409,19 +409,31 @@ func (a Actor) OnMinerWindowedPoStFailure(rt Runtime, params *OnMinerWindowedPoS
 	var claim *Claim
 	var st State
 	rt.State().Transaction(&st, func() interface{} {
+		faulty, err := st.hasDetectedFault(adt.AsStore(rt), minerAddr)
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "Failed to check if miner was faulty already: %s", err)
+		}
+
 		if err := st.putDetectedFault(adt.AsStore(rt), minerAddr); err != nil {
 			rt.Abortf(exitcode.ErrIllegalState, "Failed to put miner fault: %v", err)
 		}
```

Comment:
We can avoid this if faulty (in fact we should otherwise we are substracting numMinersMeetingMinPower once too many in `putDetectedFault`). May be worth moving that check from #198 to your below check